model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,777 @@
+        } else {
+            UserSessionModel userSession = getUserSession(realm, id);
+            userSessionEntityStream = userSession != null ? Stream.of(userSession) : Stream.empty();
+        }
+
+        return userSessionEntityStream
+                .filter(predicate)
+                .findFirst()
+                .orElse(null);
+    }
+
+    @Override
+    public long getActiveUserSessions(RealmModel realm, ClientModel client) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, false)
+                .compare(UserSessionModel.SearchableFields.CLIENT_ID, ModelCriteriaBuilder.Operator.EQ, client.getId());
+
+        LOG.tracef("getActiveUserSessions(%s, %s)%s", realm, client, getShortStackTrace());
+
+        return userSessionTx.getCount(mcb);
+    }
+
+    @Override
+    public Map<String, Long> getActiveClientSessionStats(RealmModel realm, boolean offline) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, offline);
+
+        LOG.tracef("getActiveClientSessionStats(%s, %s)%s", realm, offline, getShortStackTrace());
+
+        return userSessionTx.getUpdatedNotRemoved(mcb)
+                .map(MapUserSessionEntity::getAuthenticatedClientSessions)
+                .map(Map::keySet)
+                .flatMap(Collection::stream)
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+    }
+
+    @Override
+    public void removeUserSession(RealmModel realm, UserSessionModel session) {
+        Objects.requireNonNull(session, "The provided user session can't be null!");
+
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, false)
+                .compare(UserSessionModel.SearchableFields.ID, ModelCriteriaBuilder.Operator.EQ, UUID.fromString(session.getId()));
+
+        LOG.tracef("removeUserSession(%s, %s)%s", realm, session, getShortStackTrace());
+
+        userSessionTx.getUpdatedNotRemoved(mcb).forEach(userSession -> {
+            userSession.getAuthenticatedClientSessions().values().stream().forEach(clientSessionTx::delete);
+            userSessionTx.delete(userSession.getId());
+        });
+
+    }
+
+    @Override
+    public void removeUserSessions(RealmModel realm, UserModel user) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = userSessionStore.createCriteriaBuilder()
+                .compare(UserSessionModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserSessionModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, user.getId());
+
+        LOG.tracef("removeUserSessions(%s, %s)%s", realm, user, getShortStackTrace());
+
+        userSessionTx.getUpdatedNotRemoved(mcb).forEach(userSession -> {
+            userSession.getAuthenticatedClientSessions().values().stream().forEach(clientSessionTx::delete);
+            userSessionTx.delete(userSession.getId());
+        });
```

Comment:
Can this be optimized to leverage `tx.delete(mcb)`? e.g. by obtaining list of user sessions `lus` first, then performing two removals:
- deletion of client sessions where user_session_id in `lus`
- then performing `userSessionTx.delete(mcb)`